containerize/command.go

```diff
@@ -0,0 +1,261 @@
+			syscall.CLONE_NEWPID | // Isolated PID namespace
+			syscall.CLONE_NEWNET, // Isolated network environment
+
+		// This requires more magic if we want to really chroot somewhere (by itself it does not do much)
+		// syscall.CLONE_NEWNS | // Isolated mount namespace
+
+		// syscall.CLONE_NEWUSER | // Isolated user namespace
+		// Whatever uid/gid we use for the server will be mapped into root within the container
+		// UidMappings: []syscall.SysProcIDMap{
+		// 	{
+		// 		ContainerID: 0,
+		// 		HostID:      os.Getuid(),
+		// 		Size:        1,
+		// 	},
+		// },
+		// GidMappings: []syscall.SysProcIDMap{
+		// 	{
+		// 		ContainerID: 0,
+		// 		HostID:      os.Getgid(),
+		// 		Size:        1,
+		// 	},
+		// },
+	}
+
+	return &attr
+}
+
+//-------------------------------------------------------------------------------------------------
+// Returns true if the command is still running
+func (c *Command) Running() bool {
+	c.commandMutex.RLock()
+	running := c.running
+	c.commandMutex.RUnlock()
+	return running
+}
+
+// Waits for the process to finish (without relying on waitpid, which destroys process exit status info)
+func (c *Command) Wait() {
+	for c.Running() {
+		time.Sleep(100 * time.Millisecond)
+	}
+}
+
+//-------------------------------------------------------------------------------------------------
+// Terminates the command if it is running (including all sub-processes)
+func (c *Command) Kill() {
+	if c.Running() {
+		c.commandMutex.Lock()
+		defer c.commandMutex.Unlock()
+
+		// Kill the whole process group
+		killPg(c.pid())
+	}
+}
+
+func killPg(pgid int) error {
+	if pgid > 0 {
+		pgid = -pgid
+	}
+
+	return syscall.Kill(pgid, syscall.SIGKILL)
+}
+
+//-------------------------------------------------------------------------------------------------
+func (c *Command) ResultCode() (int, error) {
+	if c.Running() {
+		return 0, errors.New("no result code available for a running command")
+	}
+
+	c.commandMutex.RLock()
+	defer c.commandMutex.RUnlock()
+	return c.executor.ProcessState.ExitCode(), nil
+}
+
+func (c *Command) ResultDescription() (string, error) {
+	if c.Running() {
+		return "", errors.New("no result description available for a running command")
+	}
+
+	c.commandMutex.RLock()
+	defer c.commandMutex.RUnlock()
+	return c.executor.ProcessState.String(), nil
+}
+
+//-------------------------------------------------------------------------------------------------
+func (c *Command) Failure() error {
+	c.commandMutex.RLock()
+	defer c.commandMutex.RUnlock()
+	return c.failure
+}
+
+//-------------------------------------------------------------------------------------------------
+// Closes log files, releases other resources used by the command
+func (c *Command) Close() {
+	// Make sure the process has stopped
+	if c.Running() {
+		c.Kill()
+	}
+
+	// Make sure we release all the resources associated with the command
+	c.Wait()
+
+	// Close the log stream
+	c.log.Close()
```

Comment:
These operations can fail, right? Should you be returning an error?

Reply:
Good call, added